lib/experimental/custom-sources/meta.php

```diff
@@ -0,0 +1,36 @@
+<?php
+/**
+ * Meta Custom Source
+ *
+ * @package gutenberg
+ */
+
+return array(
+	'name'         => 'meta',
+	'apply_source' => function ( $block_content, $block_instance, $meta_field, $attribute_config ) {
+		// We should probably also check if the meta field exists but for now it's okay because
+		// if it doesn't, `get_post_meta()` will just return an empty string.
+		$meta_value = get_post_meta( $block_instance->context['postId'], $meta_field, true );
+		$p          = new WP_HTML_Tag_Processor( $block_content );
+		$found      = $p->next_tag(
+			array(
+				// TODO: In the future, when blocks other than Paragraph and Image are
+				// supported, we should build the full query from CSS selector.
+				'tag_name' => $attribute_config['selector'],
+			)
+		);
+		if ( ! $found ) {
+			return $block_content;
+		}
+		$tag_name = $p->get_tag();
+		$markup   = "<$tag_name>$meta_value</$tag_name>";
+		$p2       = new WP_HTML_Tag_Processor( $markup );
+		$p2->next_tag();
+		$names = $p->get_attribute_names_with_prefix( '' );
+		foreach ( $names as $name ) {
+			$p2->set_attribute( $name, $p->get_attribute( $name ) );
+		}
+
+		return $p2 . '';
```

Comment:
That's a good catch! I've just double checked and we don't need the `get_updated_html()` because `WP_HTML_Tag_Processor::__toString()` calls `get_updated_html()` automatically ([link](https://developer.wordpress.org/reference/classes/wp_html_tag_processor/__tostring/#source)).

I don't know if we need `apply_attributes_updates()` because I don't understand its use case. 

@dmsnell Could you enlighten me on what `apply_attributes_updates()` does? Is it always necessary to call it after calling `p->set_attribute()`? Or it only necessary in some circumstances?

Reply:
userspace should never worry about calling `apply_attribute_updates()`. that should be an internal-use only function

Reply:
gotcha. Thanks @dmsnell !